fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentTarget.java

```diff
@@ -0,0 +1,206 @@
+	 * @return the attached data, initialized if originally absent
+	 */
+	default <A> A getAttachedOrCreate(AttachmentType<A> type, Supplier<A> initializer) {
+		A attached = getAttached(type);
+
+		if (attached != null) {
+			return attached;
+		} else {
+			A initialized = Objects.requireNonNull(initializer.get(), "initializer result cannot be null");
+			setAttached(type, initialized);
+			return initialized;
+		}
+	}
+
+	/**
+	 * Specialization of {@link #getAttachedOrCreate(AttachmentType, Supplier)}, but <i>only for attachment types with
+	 * {@link AttachmentType#initializer() initializers}.</i> It will throw an exception if one is not present.
+	 *
+	 * @param type the attachment type
+	 * @param <A>  the type of the data
+	 * @return the attached data, initialized if originally absent
+	 */
+	default <A> A getAttachedOrCreate(AttachmentType<A> type) {
+		Supplier<A> init = type.initializer();
+
+		if (init == null) {
+			throw new IllegalArgumentException("Single-argument getAttachedOrCreate is reserved for attachment types with default initializers");
+		}
+
+		return getAttachedOrCreate(type, init);
+	}
+
+	/**
+	 * Gets the data associated with the given {@link AttachmentType}, or returns the provided default value if it doesn't exist.
+	 * Unlike {@link #getAttachedOrCreate(AttachmentType, Supplier)}, this doesn't initialize the attachment with the default value.
+	 *
+	 * @param type         the attachment type
+	 * @param defaultValue the default value to use as fallback
+	 * @param <A>          the type of the attached data
+	 * @return the attached data, or the default value
+	 */
+	default <A> A getAttachedOrElse(AttachmentType<A> type, A defaultValue) {
+		Objects.requireNonNull(defaultValue, "default value cannot be null");
+
+		A attached = getAttached(type);
+		return attached == null ? defaultValue : attached;
+	}
+
+	/**
+	 * Gets the data associated with the given {@link AttachmentType}, or gets the provided default value from the
+	 * argument if it doesn't exist. Unlike {@link #getAttachedOrCreate(AttachmentType, Supplier)}, this doesn't
+	 * initialize the attachment with the default value.
+	 *
+	 * @param type         the attachment type
+	 * @param defaultValue the default value supplier to use as fallback
+	 * @param <A>          the type of the attached data
+	 * @return the attached data, or the default value
+	 */
+	default <A> A getAttachedOrGet(AttachmentType<A> type, Supplier<A> defaultValue) {
+		Objects.requireNonNull(defaultValue, "default value cannot be null");
+
+		A attached = getAttached(type);
+		return attached == null ? defaultValue.get() : attached;
+	}
+
+	/**
+	 * Sets the data associated with the given {@link AttachmentType}. Passing {@code null} removes the data.
+	 *
+	 * @param type  the attachment type
+	 * @param value the new value
+	 * @param <A>   the type of the data
+	 * @return the previous data
+	 */
+	@Nullable
+	default <A> A setAttached(AttachmentType<A> type, @Nullable A value) {
```

Comment:
The data will not be saved automatically for block entities and chunks. Any change to attached data for BEs or chunks needs to mark the target for saving (`setChanged()` and `setUnsaved(true)` respectively). Additionally, `setAttached` should be overriden by chunks and block entities to call `setChanged` and `setUnsaved` as appropriate.

From the NF docs:
![image](https://github.com/FabricMC/fabric/assets/13494793/e4a2efef-72f0-45c7-b4fc-79092eab8bd4)